query/query.go

```diff
@@ -0,0 +1,169 @@
+	ListByCreator(ctx context.Context, spaceID uuid.UUID, creatorID uuid.UUID) ([]Query, error)
+	Load(ctx context.Context, queryID uuid.UUID, spaceID uuid.UUID) (*Query, error)
+	Delete(ctx context.Context, ID uuid.UUID) error
+}
+
+// NewQueryRepository creates a new storage type.
+func NewQueryRepository(db *gorm.DB) Repository {
+	return &GormQueryRepository{db: db}
+}
+
+// GormQueryRepository is the implementation of the storage interface for Labels.
+type GormQueryRepository struct {
+	db *gorm.DB
+}
+
+// GetETagData returns the field values to use to generate the ETag
+func (q Query) GetETagData() []interface{} {
+	return []interface{}{q.ID, strconv.FormatInt(q.UpdatedAt.Unix(), 10)}
+}
+
+// QueryTableName constant that holds table name of Labels
+const QueryTableName = "queries"
+
+// Create a new label
+func (r *GormQueryRepository) Create(ctx context.Context, q *Query) error {
+	defer goa.MeasureSince([]string{"goa", "db", "Query", "create"}, time.Now())
+	q.ID = uuid.NewV4()
+	if strings.TrimSpace(q.Title) == "" {
+		return errors.NewBadParameterError("query name cannot be empty string", q.Title).Expected("non empty string")
+	}
+	if q.Creator == uuid.Nil {
+		return errors.NewBadParameterError("creator cannot be nil", q.Creator).Expected("valid user ID")
+	}
+	err := r.db.Create(q).Error
+	if err != nil {
+		// combination of name and space ID should be unique
+		if gormsupport.IsUniqueViolation(err, "queries_title_space_id_creator_unique") {
+			log.Error(ctx, map[string]interface{}{
+				"err":      err,
+				"title":    q.Title,
+				"space_id": q.SpaceID,
+			}, "unable to create query because a query with same title already exists in the space by same creator")
+			return errors.NewDataConflictError(fmt.Sprintf("query already exists with title = %s , space_id = %s, creator = %s", q.Title, q.SpaceID, q.Creator))
+		}
+		log.Error(ctx, map[string]interface{}{}, "error adding Query: %s", err.Error())
+		return err
+	}
+	return nil
+}
+
+// List all labels in a space
+func (r *GormQueryRepository) List(ctx context.Context, spaceID uuid.UUID) ([]Query, error) {
+	defer goa.MeasureSince([]string{"goa", "db", "Query", "list"}, time.Now())
+	var objs []Query
+	err := r.db.Where("space_id = ?", spaceID).Find(&objs).Error
+	if err != nil && err != gorm.ErrRecordNotFound {
+		return nil, err
+	}
+	return objs, nil
+}
+
+// ListByCreator all labels in a space by a creator
+func (r *GormQueryRepository) ListByCreator(ctx context.Context, spaceID uuid.UUID, creatorID uuid.UUID) ([]Query, error) {
+	defer goa.MeasureSince([]string{"goa", "db", "Query", "list"}, time.Now())
+	var objs []Query
+	err := r.db.Where("space_id = ? AND creator=?", spaceID, creatorID).Find(&objs).Error
+	if err != nil && err != gorm.ErrRecordNotFound {
+		return nil, err
+	}
+	return objs, nil
+}
+
+// Load Query in a space
+func (r *GormQueryRepository) Load(ctx context.Context, ID uuid.UUID, spaceID uuid.UUID) (*Query, error) {
```

Comment:
Why do you need the space ID in order to load a query? Is the query's ID not "unique enough"? ;)

Reply:
hmm yes, I tried to align with other APIs hence made it space aware. it is really not required.
Will you suggest to replace spaceID by creatorID and add authentication to this API? user must be logged in and only he can load this particular object?

Reply:
This is the repo level. There should be no authentication. But the controller should double check that once loaded, the  query's author matches the current authenticated user.